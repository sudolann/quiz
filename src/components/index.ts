export * from './errorMessage/ErrorMessage';
export * from './loadingBar/LoadingBar';
export * from './questionCard/QuestionCard';
export * from './quizForm/QuizForm';
export * from './results/Results';


