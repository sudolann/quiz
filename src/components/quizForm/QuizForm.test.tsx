import { shallow } from 'enzyme';
import React from 'react';
import { QuizForm, ErrorMessage, LoadingBar, QuestionCard } from '../../components/';
import renderer from 'react-test-renderer';
import * as redux from 'react-redux';
import { act } from '@testing-library/react';

const mockDispatch = jest.fn();
const spy = jest.spyOn(redux, 'useSelector');
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: (): jest.Mock<any, any> => mockDispatch,
}));

describe('component', (): void => {
  describe('QuizForm', (): void => {
    it('match snapshot', (): void => {
      spy.mockReturnValue({
        list: [],
        loading: false,
      });
      const tree = renderer.create(<QuizForm />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders header', (): void => {
      const wrapper = shallow(<QuizForm />);
      const h1 = wrapper.find('h1');
      expect(h1.text()).toBe('Welcome to the Trivia Challenge!');
    });
    it('renders disabled button by default', (): void => {
      const wrapper = shallow(<QuizForm />);
      const button = wrapper.find('Button[id="quiz-btn"]');
      expect(button.prop('disabled')).toBeTruthy();
    });
    it('renders select input without selected value', (): void => {
      const wrapper = shallow(<QuizForm />);
      const selectInput = wrapper.find('select');
      expect(selectInput.prop('value')).toBe('select');
    });
    it('renders select input number without value', (): void => {
      const wrapper = shallow(<QuizForm />);
      const inputNumber = wrapper.find('input[placeholder="amount"]');
      expect(inputNumber.prop('value')).toBe(undefined);
    });
    it('call dispatch upon clicking button', async () => {
      const wrapper = shallow(<QuizForm />);
      wrapper.find('select').simulate('change', { target: { name: 'difficulty', value: 'easy' } });
      expect(wrapper.find('select').props().value).toBe('easy');
      wrapper.find('input[placeholder="amount"]').simulate('change', { target: { name: 'amount', value: 2 } });
      expect(wrapper.find('input[placeholder="amount"]').props().value).toBe(2);
      const button = wrapper.find('Button[id="quiz-btn"]');
      wrapper.update();
      expect(button.prop('disabled')).toBeFalsy();
      const mockEvent = {
        preventDefault: jest.fn(),
      };

      act(() => {
        wrapper.find('.form').prop<Function>('onSubmit')(mockEvent);
      });
      wrapper.update();
      expect(mockDispatch).toHaveBeenCalledTimes(1);
    });

    it('renders error message', (): void => {
      spy.mockClear();
      spy.mockReturnValue({
        list: [],
        loading: false,
        error: 'test message',
      });
      const wrapper = shallow(<QuizForm />);
      expect(wrapper.containsMatchingElement(<ErrorMessage message='test message' />));
    });
    it('renders loading component', (): void => {
      spy.mockClear();
      spy.mockReturnValue({
        list: [],
        loading: true,
      });
      const wrapper = shallow(<QuizForm />);
      expect(wrapper.containsMatchingElement(<LoadingBar />));
    });

    it('renders QuestionCard component', (): void => {
      spy.mockClear();
      spy.mockReturnValue({
        list: [
          {
            category: 'General Knowledge',
            question: 'Nutella is produced by the German company Ferrero.',
            correct_answer: 'False',
          },
          { category: 'General Knowledge', question: 'Romanian belongs to the Romance language family, shared with French, Spanish, Portuguese and Italian. ', correct_answer: 'True' },
        ],
        loading: false,
      });
      const wrapper = shallow(<QuizForm />);
      expect(wrapper.containsMatchingElement(<QuestionCard />));
    });
  });
});
