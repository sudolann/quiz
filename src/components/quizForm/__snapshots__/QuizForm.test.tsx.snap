// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component QuizForm match snapshot 1`] = `
<div
  className="container"
>
  <header
    className="header"
  >
    <h1>
      Welcome to the Trivia 
      <br />
      Challenge!
    </h1>
  </header>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="wrapper"
    >
      <select
        className="form__input form__input--select"
        name="difficulty"
        onChange={[Function]}
        placeholder="difficulty"
        value="select"
      >
        <option
          disabled={true}
          value="select"
        >
          difficulty
        </option>
        <option
          value="easy"
        >
          easy
        </option>
        <option
          value="hard"
        >
          hard
        </option>
      </select>
      <input
        className="form__input form__input--number"
        name="amount"
        onChange={[Function]}
        placeholder="amount"
        type="number"
      />
    </div>
    <button
      className="ant-btn ant-btn-primary"
      disabled={true}
      id="quiz-btn"
      onClick={[Function]}
      type="submit"
    >
      <span>
        begin
      </span>
    </button>
  </form>
</div>
`;
